usr/src/man/man1/mesg.1
.\"
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\"
.\" The Institute of Electrical and Electronics Engineers and The Open
.\" Group, have given us permission to reprint portions of their
.\" documentation.
.\"
.\" In the following statement, the phrase ``this text'' refers to portions
.\" of the system documentation.
.\"
.\" Portions of this text are reprinted and reproduced in electronic form
.\" in the SunOS Reference Manual, from IEEE Std 1003.1, 2004 Edition,
.\" Standard for Information Technology -- Portable Operating System
.\" Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group.  In the event of any discrepancy
.\" between these versions and the original IEEE and The Open Group
.\" Standard, the original IEEE and The Open Group Standard is the referee
.\" document.  The original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"
.\" This notice shall appear on any product containing this material.
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright 1989 AT&T
.\" Copyright (c) 1992, X/Open Company Limited.  All Rights Reserved.
.\" Portions Copyright (c) 1995, Sun Microsystems, Inc.  All Rights Reserved
.\"
.TH MESG 1 "Oct 31, 1997"
.SH NAME
mesg \- permit or deny messages
.SH SYNOPSIS
.LP
.nf
\fBmesg\fR [\fB-n\fR | \fB-y\fR | n | y]
.fi

.SH DESCRIPTION
.sp
.LP
The \fBmesg\fR utility will control whether other users are allowed to send
messages via \fBwrite\fR(1), \fBtalk\fR(1), or other utilities to a terminal
device. The terminal device affected is determined by searching for the first
terminal in the sequence of devices associated with standard input, standard
output, and standard error, respectively. With no arguments, \fBmesg\fR reports
the current state without changing it. Processes with appropriate privileges
may be able to send messages to the terminal independent of the current state.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-n\fR|\fBn\fR \fR
.ad
.RS 9n
Denies permission to other users to send message to the terminal. See
\fBwrite\fR(1).
.RE

.sp
.ne 2
.na
\fB\fB-y\fR|\fBy\fR \fR
.ad
.RS 9n
Grants permission to other users to send messages to the terminal.
.RE

.SH ENVIRONMENT VARIABLES
.sp
.LP
See \fBenviron\fR(7) for descriptions of the following environment variables
that affect the execution of \fBmesg\fR: \fBLANG\fR, \fBLC_ALL\fR,
\fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR \fR
.ad
.RS 6n
if messages are receivable.
.RE

.sp
.ne 2
.na
\fB\fB1\fR \fR
.ad
.RS 6n
if messages are not receivable.
.RE

.sp
.ne 2
.na
\fB\fB2\fR \fR
.ad
.RS 6n
on error.
.RE

.SH FILES
.sp
.ne 2
.na
\fB\fB/dev/tty*\fR \fR
.ad
.RS 15n
terminal devices
.RE

.sp
.ne 2
.na
\fB\fB/dev/pts/*\fR \fR
.ad
.RS 15n
terminal devices
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(7) for descriptions of the following attributes:
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
Interface Stability	Standard
.TE

.SH SEE ALSO
.sp
.LP
.BR talk (1),
.BR write (1),
.BR attributes (7),
.BR environ (7),
.BR standards (7)


usr/src/cmd/mesg/mesg.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2003 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

#pragma ident	"%Z%%M%	%I%	%E% SMI"

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


/*
 * mesg -- set current tty to accept or
 *	forbid write permission.
 *
 *	mesg [-y | -n | y | n]
 *		y allow messages
 *		n forbid messages
 *	return codes
 *		0 if messages are ON or turned ON
 *		1 if messages are OFF or turned OFF
 *		2 if an error occurs
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <locale.h>
#include <libintl.h>
#include <sys/types.h>
#include <sys/stat.h>

static void error(const char *s);
static void newmode(mode_t m);
static void usage(void);

static char *tty;

int
main(int argc, char *argv[])
{
	int i, c, r = 0;
	int action = 0;
	struct stat sbuf;

	extern int optind;

	(void) setlocale(LC_ALL, "");
#if !defined(TEXT_DOMAIN)		/* Should be defined by cc -D */
#define	TEXT_DOMAIN	"SYS_TEST"	/* Use this only if it weren't */
#endif
	(void) textdomain(TEXT_DOMAIN);

	/*
	 * Check stdin, stdout and stderr, in order, for a tty
	 */
	for (i = 0; i <= 2; i++) {
		if ((tty = ttyname(i)) != NULL)
			break;
	}

	if (stat(tty, &sbuf) < 0)
		error("cannot stat");

	if (argc < 2) {
		if (sbuf.st_mode & (S_IWGRP | S_IWOTH)) {
			(void) printf("is y\n");
		} else {
			r = 1;
			(void) printf("is n\n");
		}
		exit(r);
	}

	while ((c = getopt(argc, argv, "yn")) != EOF) {
		switch (c) {
		case 'y':
			if (action > 0)
				usage();

			newmode(S_IRUSR | S_IWUSR | S_IWGRP);
			action++;
			break;

		case 'n':
			if (action > 0)
				usage();

			newmode(S_IRUSR | S_IWUSR);
			r = 1;
			action++;
			break;

		case '?':
			usage();
			break;
		}
	}

	/*
	 * Required for POSIX.2
	 */
	if (argc > optind) {
		if (action > 0)
			usage();

		switch (*argv[optind]) {
		case 'y':
			newmode(S_IRUSR | S_IWUSR | S_IWGRP);
			break;

		case 'n':
			newmode(S_IRUSR | S_IWUSR);
			r = 1;
			break;

		default:
			usage();
			break;
		}
	}

	return (r);
}

void
error(const char *s)
{
	(void) fprintf(stderr, "mesg: ");
	(void) fprintf(stderr, "%s\n", s);
	exit(2);
}

void
newmode(mode_t m)
{
	if (chmod(tty, m) < 0)
		error("cannot change mode");
}

void
usage(void)
{
	(void) fprintf(stderr, gettext("usage: mesg [-y | -n | y | n]\n"));
	exit(2);
}
